62.CAD/fragment_shader.hlsl

```diff
@@ -15,6 +16,77 @@ void endInvocationInterlockEXT();
 // Write a general one, and maybe another one that uses precomputed values, and move these to somewhere nice in our builtin hlsl shaders if we don't have one
 // See: https://github.com/erich666/GraphicsGems/blob/master/gems/Roots3And4.c
 
+struct ArrayAccessor
+{
+    uint32_t styleIdx;
+    using value_type = float;
+
+    float operator[](const uint32_t ix)
+    {
+        return lineStyles[styleIdx].stipplePattern[ix];
+    }
+};
+
+template<typename float_t, typename CurveType, typename StyleAccessor>
+struct LineStyleClipper
+{
+    using float2_t = vector<float_t, 2>;
+
+    static LineStyleClipper<float_t, CurveType, StyleAccessor> construct(StyleAccessor styleAccessor, 
+                                                                         CurveType curve,
+                                                                         typename CurveType::ArcLenCalculator arcLenCalc)
+    {
+        LineStyleClipper<float_t, CurveType, StyleAccessor> ret = { styleAccessor, curve, arcLenCalc };
+        return ret;
+    }
+    
+    float2_t operator()(float t)
+    {
+        const LineStyle style = lineStyles[styleAccessor.styleIdx];
+        const float arcLen = arcLenCalc.calcArcLen(t);
+        t = clamp(t, 0.0f, 1.0f);
+        const float worldSpaceArcLen = arcLen * float(globals.worldToScreenRatio);
+        float_t normalizedPlaceInPattern = frac(worldSpaceArcLen * style.reciprocalStipplePatternLen + style.phaseShift);
+        uint32_t patternIdx = nbl::hlsl::upper_bound(styleAccessor, 0, style.stipplePatternSize, normalizedPlaceInPattern);
+        
+        // odd patternIdx means a "no draw section" and current candidate should split into two nearest draw sections
+        if(patternIdx & 0x1)
+        {   
+            float diffToLeftDrawableSection = (patternIdx == 0) ? 0.0f : style.stipplePattern[patternIdx-1];
```

Comment:
btw patternIdx won't be 0 cause you made sure it's odd by &0x1

Reply:
done